### PR TITLE
Resolve indefinite hang during  mesheryctl system login with an active Meshery Cloud session

### LINKED ISSUE #8247: [mesheryctl] mesheryctl system login hangs indefinitely when meshery cloud is already logged in
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior

When `mesheryctl system login` is run, a page opens in the browser asking the user to log into Meshery cloud. Once the user logs in, he is given a successful login message.

However if the user is already logged into Meshery cloud in the browser, they are shown the Meshery cloud dashboard and the command hangs indefinitely in the terminal.

#### Desired Behavior

A user should be able to use `mesheryctl system login` to login even if they are already logged into Meshery cloud in the browser

#### Screenshots/Logs

https://github.com/meshery/meshery/assets/76054921/c13f0f40-ab90-4aa8-8862-cd37929b7f4a

---

#### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 📚 [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- 📺 [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- ⌨️ [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- ⌨️ [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by yash37158**
**Notes for Reviewers**

This PR fixes #8247 
Fixed the indefinite hang during `mesheryctl system login` with an active Meshery Cloud session

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by yash37158** [comment]
here is the reference:
[Click to watch the video](https://youtu.be/2PXeEp4LrWg)

**Comment by yash37158** [comment]
// @abdullah1308 @Chadha93 @leecalcote

**Comment by leecalcote** [comment]
Thanks for posting a video.

**Comment by leecalcote** [comment]
@zakisk review?

**Comment by yash37158** [comment]
You are always welcome @leecalcote, I also mentioned a bug in meshery cloud on Slack if you can spread some light on that too

**Comment by zakisk** [comment]
> @zakisk review?

Yes

**Comment by leecalcote** [comment]
@zakisk [this failure](https://github.com/meshery/meshery/actions/runs/5695534531/job/15438880533?pr=8299#step:5:12) is unrelated to this PR. Is this something that you can look into?

**Comment by yash37158** [comment]
Hey @leecalcote, @Chadha93, @abdullah1308 are we good to merge this?

**Comment by zakisk** [comment]
> @zakisk [this failure](https://github.com/meshery/meshery/actions/runs/5695534531/job/15438880533?pr=8299#step:5:12) is unrelated to this PR. Is this something that you can look into?

@leecalcote this error error is same as we've got [here](https://github.com/meshery/meshery/pull/8272#issuecomment-1650064762) it as well, I am finding its solution. it wiil may be a `good first issue` or could be complex if it is used in project on many places and needs to be changed on every place with precaution.

**Comment by zakisk** [comment]
@leecalcote I've checked and I think that it can be a `good first issue` because [json.RawMessage](https://pkg.go.dev/encoding/json#RawMessage) is nothing but a wrapper to `[]byte` type. it is used in project on many places but as json.RawMessage's underlying type is []byte it wouldn't cause any problem, what's your take about this?

**Comment by vishalvivekm** [comment]
@yash37158 Let's discuss this on Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

**Comment by leecalcote** [comment]
> @leecalcote I've checked and I think that it can be a `good first issue` because [json.RawMessage](https://pkg.go.dev/encoding/json#RawMessage) is nothing but a wrapper to `[]byte` type. it is used in project on many places but as json.RawMessage's underlying type is []byte it wouldn't cause any problem, what's your take about this?

It'd be good to hear from @hexxdump

**Comment by hexxdump** [comment]
> > @leecalcote I've checked and I think that it can be a `good first issue` because [json.RawMessage](https://pkg.go.dev/encoding/json#RawMessage) is nothing but a wrapper to `[]byte` type. it is used in project on many places but as json.RawMessage's underlying type is []byte it wouldn't cause any problem, what's your take about this?
> 
> It'd be good to hear from @hexxdump

Clearly, RawMessage has advantages over plain []byte, however we are introducing a new package (encoding/json) dependency here. 
Before deciding whether to choose RawMessage or not, it would be good to root cause why the parsing failed in [here](https://github.com/meshery/meshery/actions/runs/5695534531/job/15438880533?pr=8299#step:5:12) and [here](https://github.com/meshery/meshery/actions/runs/5657521386/job/15326944619?pr=8272#step:5:31)

**Comment by yash37158** [comment]
Hey @leecalcote @suhail34 @abdullah1308  reminder for this PR

**Review by suhail34 — CHANGES_REQUESTED** [review]
I wasn't having auth.json file and was logged in to meshery cloud this what auth.json looked like after running `mesheryctl system login`
image

**Comment by suhail34** [comment]
Will debug why does this happened

File: mesheryctl/pkg/utils/auth.go [hunk 1]
```diff
@@ -342,18 +343,41 @@ func initiateRemoteProviderAuth(provider Provider) (string, error) {
...
-		return token, err
+	// Pause until we get the response on the channel or timeout occurs
+	select {
+	case token := <-tokenChan:
+		// Shut down the server
+		if err := srv.Shutdown(context.Background()); err != nil {
+			return "", err
+		}
+		return token, nil
+	case <-time.After(5 * time.Second): // Set the timeout duration as needed
```
Comment (abdullah1308) [inline mesheryctl/pkg/utils/auth.go:361]: @yash37158 This does not solve the issue. The issue required that `mesheryctl system login` be successful when an active meshery cloud session was already present in the browser.

Also this line causes the bug where `mesheryctl system login` finishes execution before the login is completed on the following page. This is invalid behaviour.

image

**Comment by abdullah1308** [comment]
Closing this PR for 2 reasons -

1. The issue has not been addressed 
2. The changes made introduce bugs in `mesheryctl system login` -
     a. It can take upto 10-15 seconds to log in through SSO. If the log in isn't completed within 5 seconds, the changes make running the command useless causing the login to fail.

@yash37158 Please feel free to open another PR.

**Comment by yash37158** [comment]
Thanks for your feedback @abdullah1308, will come with a valid PR again

**Pull request closed without merging**